### ISSUE TITLE
`new_without_default` doesn't trigger on `const fn`

### ISSUE THREAD

**Issue by robertbastian** [rust-lang/rust-clippy#10877]
### Summary

Triggers on `pub fn new() -> Self` but not `pub const fn new() -> Self`

### Lint Name

new_without_default

### Reproducer

I tried this code:

```rust
pub struct Foo;

impl Foo {
  pub const fn new() -> {
    Self
  }
}
```

I expected to see this happen: lint triggers

Instead, this happened: lint doesn't trigger

### Version

```text
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=256edf22e6afcae3a1f23322a19851de
```

**Comment by Centri3** [comment]
`Default::default` is not const so afaict this is correct behavior, same with `unsafe fn new`

**Comment by robertbastian** [comment]
`Default::default` can call `const fn new` though. It can't call `unsafe fn` new without being unsafe itself, so that's different.

**Comment by Centri3** [comment]
Yeah that's fair actually, this is still intentionally done by the lint's author but this should probably be changed at some point. @rustbot claim
